dcurve/src/internal/fetch.ts

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+import { Contract } from 'ethers';
+import { Contribution, GrantRoundContributions, GrantRoundFetchArgs } from '../../src/types';
+import { abi as GRANT_REGISTRY_ABI } from '@dgrants/contracts/artifacts/contracts/GrantRegistry.sol/GrantRegistry.json';
+import { abi as GRANT_ROUND_ABI } from '@dgrants/contracts/artifacts/contracts/GrantRound.sol/GrantRound.json';
+import { abi as GRANT_ROUND_MANAGER_ABI } from '@dgrants/contracts/artifacts/contracts/GrantRoundManager.sol/GrantRoundManager.json';
+import { formatUnits } from 'ethers/lib/utils';
+
+const ERC20_ABI = ['function balanceOf(address) view returns (uint)'];
+
+/**
+ * Returns GrantRoundContributions describing the round and any contributions
+ * @param {GrantRoundFetchArgs}
+ */
+export const fetch = async (args: GrantRoundFetchArgs) => {
+  // fetch the grantList
+  const registry = new Contract(args.grantRegistry, GRANT_REGISTRY_ABI, args.provider);
+  // fetch the grantRound
+  const round = new Contract(args.grantRound, GRANT_ROUND_ABI, args.provider);
+  // fetch grantRoundManager
+  const roundManager = new Contract(args.grantRoundManager, GRANT_ROUND_MANAGER_ABI, args.provider);
+
+  // set-up
+  const [allGrants, matchingToken, donationToken, grantDonations] = await Promise.all([
+    registry.getAllGrants(), // TODO: https://github.com/dcgtc/dgrants/pull/91#discussion_r692567688
+    round.matchingToken(),
+    roundManager.donationToken(),
+    roundManager.queryFilter(roundManager.filters.GrantDonation()), // TODO: https://github.com/dcgtc/dgrants/pull/91#discussion_r692569977
+  ]);
+
+  // collect the grants into a grantId->payoutAddress obj
+  const grantsDict = allGrants.reduce((grants: Record<string, string>, grant: Record<string, string>, key: number) => {
+    grants[key.toString()] = grant.payee;
+
+    return grants;
+  }, {});
+
+  // get token info
+  const matchingTokenContract = new Contract(matchingToken, ERC20_ABI, args.provider);
+
+  // fetch & ignore Contributions
+  const contributions: Contribution[] = [];
+
+  // records matching contributions in the expected format (Contribution[])
+  grantDonations.forEach(async (contribution) => {
+    // get tx details to pull contributor details from
+    const tx = await contribution.getTransaction();
+
+    // check that the contribution is valid
+    const grantId = contribution?.args?.grantId.toString();
+    const inRound = contribution?.args?.rounds.includes(args.grantRound);
+    const isIgnoredGrant = args?.ignore?.grants?.includes(grantId);
+    const isIgnoredContributor = args?.ignore?.contributionAddress?.indexOf(tx.from) !== -1;
+
+    // only include transactions from this grantRound which havent been ignored
+    if (inRound && !isIgnoredGrant && !isIgnoredContributor) {
+      contributions.push({
+        grantId: grantId,
+        amount: parseFloat(
+          formatUnits(contribution?.args?.donationAmount, args.supportedTokens[donationToken].decimals)
+        ),
```

Comment:
Can we use `formatUnits(...).toNumber()` here instead? Only because `toNumber()` should be more robust since it's part of the class instance and will throw if it needs to, whereas `parseFloat` is more likely result in unexpected behavior/NaNs for some edge cases

Reply:
@mds1 `formatUnits` returns a `string` ! We could do a `Number(...)`

Reply:
Ah right I was thinking it returned a BigNumber. Hmm, I think `BigNumber.from(stuff).toNumber()` would be best at helping us catch any sneaky bugs/bad values that made it this far

Reply:
`BigNumber` will only accept integers and `formatUnits` will always return a decimal string with at least one decimal - the biggest number we can safely parse here would be `9007199254740991` - if we think we'll have to deal with any numbers larger than that then we may have to refactor to use `BigNumber` throughout - what do you think @mds1 @thelostone-mc?

Reply:
Ahhh yes yes ok, we can stick with parseFloat for now, sorry! Since we're dealing with regular numbers we likely won't exceed that `2^53-1` limit so it's probably fine! I was just trying to think of edge cases/be cautious

Reply:
Yeah, its best to be cautious - lets see how we go - we know that in this first round we'll be in control of the matching currency and the donation currencies, and I don't think we expect to have a match or a contribution that exceeds 9 quadrillion - so we should be safe!